### PR TITLE
adding custom changes to grammes fork

### PR DESCRIPTION
**Pull request by waleedakramkhan**
https://k6scope.atlassian.net/browse/BLUEP-175

This PR adds custom changes made in https://gitlab.com/kaleidoscope.root/grammes-gremlin to this fork

### CODE REVIEW

**Comment by waleedakramkhan** [comment]
i've also added test for a custom `StringStep` functionality added by @MDanialSaleem please review

File: gremconnect/response.go
Comment (waleedakramkhan) [inline gremconnect/response.go]: Changes in this file were removed [here](https://gitlab.com/kaleidoscope.root/grammes-gremlin/-/commit/58aecaa0988a60f068717bef400c984533914001)

should be put back imo

File: gremconnect/response_test.go
Comment (waleedakramkhan) [inline gremconnect/response_test.go]: Changes in this file were removed [here](https://gitlab.com/kaleidoscope.root/grammes-gremlin/-/commit/58aecaa0988a60f068717bef400c984533914001)

should be put back imo

File: gremerror/networkerror.go
Comment (waleedakramkhan) [inline gremerror/networkerror.go]: Changes in this file were removed [here](https://gitlab.com/kaleidoscope.root/grammes-gremlin/-/commit/58aecaa0988a60f068717bef400c984533914001)

should be put back imo

File: manager/addvertex.go
Comment (waleedakramkhan) [inline manager/addvertex.go:105]: don't think this change will be accepted by maintainers of this repo

File: manager/misc.go
Comment (waleedakramkhan) [inline manager/misc.go]: don't think this change will be accepted by maintainers of this repo

File: query/traversal/addproperty.go
Comment (waleedakramkhan) [inline query/traversal/addproperty.go]: i believe that this change should be reverted..it's already handled by library in a much better design after refactoring..

File: query/traversal/graph.go
Comment (waleedakramkhan) [inline query/traversal/graph.go]: this change should also be reverted..their handling has much better design, our version is too old

File: query/traversal/stringstep.go
Comment (waleedakramkhan) [inline query/traversal/stringstep.go:3]: added by @MDanialSaleem will add tests for it

File: query/traversal/stringstep_test.go
Comment (waleedakramkhan) [inline query/traversal/stringstep_test.go:29]: tests for functionality added by @MDanialSaleem

File: query/traversal/util.go
Comment (waleedakramkhan) [inline query/traversal/util.go:91]: moved the [change](https://gitlab.com/kaleidoscope.root/grammes-gremlin/-/merge_requests/5/diffs) in refactored newer version here

**Review by waleedakramkhan — COMMENTED** [review]
Our version is too old - even older than K6scope itself i believe (at least more than 4 years) i have reviewed and commented on what i believe how the two versions should be merged but i lack the context for these changes so im importing them and leaving them here for all stakeholders to decide unanimously around these. my opinions are in the comments.

my preference would be to not introduce so many changes because from what i have seen, original library provides all methods to attain the goals we are with our custom changes except for type ones (maybe those too but not able to comment confidently).

File: manager/addvertex.go
Comment (kingzbauer) [inline manager/addvertex.go:108]: Would this catch all cases of lists? In the event that the calling function passes something like `[]int`, this would not be executed.

File: model/vertex_query.go
Comment (kingzbauer) [inline model/vertex_query.go:241]: So far I've seen both `AddVertex` and this method perform this operation on properties. Should we consolidate this into a single function?

File: model/vertex_test.go
Comment (kingzbauer) [inline model/vertex_test.go:158]: We could probably include a set property with a value of type e.g `[]int` or `[]string` and `[]interface{}` to make sure they are handled correctly.

File: query/traversal/addproperty.go
Comment (kingzbauer) [inline query/traversal/addproperty.go]: Should we explicitly check for float as well?

**Review by kingzbauer — COMMENTED** [review]
Left a couple of comments

File: gremconnect/websocket.go
Comment (MDanialSaleem) [inline gremconnect/websocket.go:58]: we should make this configurable.

File: model/vertex_test.go
Comment (waleedakramkhan) [reply model/vertex_test.go:158]: Properties are a list of interface and we cannot send int as interface, can we?

File: manager/addvertex.go
Comment (waleedakramkhan) [reply manager/addvertex.go:108]: will there ever be a case like this?

File: gremconnect/websocket.go
Comment (zubair-nabi) [reply gremconnect/websocket.go:58]: Should we add another method that takes this as param as input? Otherwise we'll have to modify this signature which might not work for upstream.

Comment (MDanialSaleem) [reply gremconnect/websocket.go:58]: I think for now, we can let it be. There needs to be a more detailed investigation into why we changed what we changed, and if there is a better way to make those changes, before we can merge this to upstream.

File: go.mod
Comment (MDanialSaleem) [inline go.mod:17]: why does this package attempts to refer to an older version of itself?

**Review by MDanialSaleem — APPROVED** [review]
one minor comment otherwise lgtm

File: gremconnect/websocket.go
Comment (waleedakramkhan) [reply gremconnect/websocket.go:58]: i tried this but this needs more parameterizing than it looks so i reverted my changes. i think it's better to open up this issue on grammes and then see how it goes

File: go.mod
Comment (waleedakramkhan) [reply go.mod:17]: this was an attempt to not change module everywhere since we want it to use our changes but it failed because go has problems with moving on from her ex - i decided to create a separate [branch](https://github.com/Kaleidoscope-Inc/grammes/tree/k6scope-dev) and use this in our code

Comment (MDanialSaleem) [reply go.mod:17]: Oh I see that makes sense. That probably has to do with go's package mechanism + forking. Anyway, LGTM!

**Review by MDanialSaleem — APPROVED** [review]

**Pull request closed without merging**
